sync/sync_getter.go

```diff
@@ -3,29 +3,50 @@ package sync
 import (
 	"context"
 	"sync"
+	"sync/atomic"
 
 	"github.com/celestiaorg/go-header"
 )
 
 // syncGetter is a Getter wrapper that ensure only one Head call happens at the time
 type syncGetter[H header.Header] struct {
+	getterLk   sync.RWMutex
+	isGetterLk atomic.Bool
 	header.Getter[H]
-
-	headLk  sync.RWMutex
-	headErr error
-	head    H
 }
 
-func (se *syncGetter[H]) Head(ctx context.Context) (H, error) {
-	// the lock construction here ensures only one routine calling Head at a time
+// Lock locks the getter for single user.
+// Reports 'true' if the lock was held by the current routine.
+// Does not require unlocking on 'false'.
+func (se *syncGetter[H]) Lock() bool {
```

Comment:
That's a good suggestion, and I even called this `TryLock` initially. The reason I went against is that `TryLock` is never blocking, while this method is. 

With `TryLock` the first routine acquiring the lock returns true, and all the subsequent routines also immediately return, but with false. In this case, we have a weird mix, where the behavior for returned bool is like with `TryLock`, while the locking is like with `Lock`. Subsequent routines are actually locked instead of immediate return. Hope this makes sense.